File: Baekjoon/BinarySearch/[9702]LIS/solution.cpp
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

#define endl "\n"

vector<int> seq, lis;
int T, N, res;

int main() {
  ios::sync_with_stdio(false);
  cin.tie(nullptr);

  cin >> T;

  for (int t = 0; t < T; t++) {
    cin >> N;

    seq = vector<int>(N);
    res = 0;

    for (int i = 0; i < N; i++) {
      cin >> seq[i];
    }

    for (int i = 0; i < N; i++) {
      lis.clear();

      for (int j = i; j < N; j++) {
        int idx = lower_bound(lis.begin(), lis.end(), seq[j]) - lis.begin();

        if (idx == lis.size()) {
          lis.push_back(seq[j]);
        } else {
          lis[idx] = seq[j];
        }

        res += lis.size();
      }
    }

    cout << "Case #" << t + 1 << ": " << res << endl;
  }

  return 0;
}

File: Baekjoon/BinarySearch/[9702]LIS/solution.py
import sys

T = int(sys.stdin.readline())
lis = []
res = 0


def lower_bound(value):
    low = 0
    high = len(lis)

    while low < high:
        mid = (low + high) // 2

        if value < lis[mid]:
            high = mid
        else:
            low = mid + 1

    return high


for t in range(T):
    N = int(sys.stdin.readline())
    seq = [0 for i in range(N)]
    res = 0

    for i in range(N):
        seq[i] = int(sys.stdin.readline())

    for i in range(N):
        lis.clear()
        for j in range(i, N):
            idx = lower_bound(seq[j])

            if idx == len(lis):
                lis.append(seq[j])
            else:
                lis[idx] = seq[j]

            res += len(lis)

    sys.stdout.write("Case #%d: %d\n" % ((t+1), res))
